R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
+  }
+  if (length(nthreads) > 1L) {
+    nthreads <- head(nthreads, 1L)
+  }
+  if (is.na(nthreads) || nthreads < 1) {
+    stop("Passed invalid 'nthreads': ", nthreads)
+  }
+  if (is.numeric(nthreads)) {
+    if (floor(nthreads) != nthreads) {
+      stop("'nthreads' must be an integer.")
+    }
+  }
+  return(as.integer(nthreads))
+}
+
+check.can.use.qdm <- function(x, params) {
+  if (inherits(x, "sparseMatrix") && !inherits(x, "dgRMatrix")) {
+    return(FALSE)
+  }
+  if ("booster" %in% names(params)) {
+    if (params$booster == "gblinear") {
+      return(FALSE)
+    }
+  }
+  if ("tree_method" %in% names(params)) {
+    if (params$tree_method %in% c("exact", "approx")) {
+      return(FALSE)
+    }
+  }
+  return(TRUE)
+}
+
+process.x.and.col.args <- function(
+  x,
+  monotone_constraints,
+  interaction_constraints,
+  feature_weights,
+  lst_args
+) {
+  if (is.null(x)) {
+    stop("'x' cannot be NULL.")
+  }
+  if (inherits(x, "xgb.DMatrix")) {
+    stop("Cannot pass 'xgb.DMatrix' as 'x' to 'xgboost()'. Try 'xgb.train()' instead.")
+  }
+  supported_x_types <- c("data.frame", "matrix", "dgCMatrix", "dgRMatrix")
+  if (!inherits(x, supported_x_types)) {
+    stop(
+      "'x' must be one of the following classes: ",
+      paste(supported_x_types, collapse = ", "),
+      ". Got: ",
+      paste(class(x), collapse = ", ")
+    )
+  }
+
+  if (NROW(feature_weights)) {
+    if (is.list(feature_weights)) {
+      feature_weights <- unlist(feature_weights)
+    }
+    if (!inherits(feature_weights, c("numeric", "integer"))) {
+      stop("'feature_weights' must be a numeric vector or named list matching to columns of 'x'.")
+    }
+    if (length(feature_weights) != ncol(x)) {
+      stop(
+        "'feature_weights' does not match in length with columns of 'x' (",
+        length(feature_weights), " vs. ", ncol(x), ")."
+      )
+    }
+    if (NROW(names(feature_weights)) && NROW(colnames(x))) {
+      matched <- match(colnames(x), names(feature_weights))
+      matched <- matched[!is.na(matched)]
+      matched <- matched[!duplicated(matched)]
+      if (length(matched) > 0 && length(matched) < length(feature_weights)) {
+        stop(
+          "'feature_weights' names do not contain all columns of 'x'. Missing: ",
+          head(setdiff(colnames(x), names(feature_weights)))
+        )
+      }
+      if (length(matched)) {
+        feature_weights <- feature_weights[matched]
+      } else {
+        warning("Names of 'feature_weights' do not match with 'x'. Names will be ignored.")
+      }
+    }
+
+    lst_args$dmatrix_args$feature_weights <- unname(feature_weights)
+  }
+
+  if (NROW(monotone_constraints)) {
+
+    if (NROW(monotone_constraints) > ncol(x)) {
+      stop(
+        "'monotone_constraints' contains more entries than there are columns in 'x' (",
+        NROW(monotone_constraints), " vs. ", ncol(x), ")."
+      )
+    }
+
+    if (is.list(monotone_constraints)) {
+
+      if (!NROW(names(monotone_constraints))) {
```

Comment:
We need the names multiple times
```r
nms <- names(monotone_constraints)
```

Reply:
Retrieving the names does not generate a deep copy of them.